static/src/javascripts/lib/cmp.js

```diff
@@ -0,0 +1,89 @@
+// @flow
+import {
+    getAdConsentState,
+    thirdPartyTrackingAdConsent,
+} from 'common/modules/commercial/ad-prefs.lib';
+
+type PurposeEvent = 'functional' | 'performance' | 'advertisement';
+
+type PurposeCallback = (state: boolean | null) => void;
+
+type Purpose = {
+    state: boolean | null,
+    callbacks: Array<PurposeCallback>,
+};
+
+let cmpIsReady = false;
+
+const purposes: { [PurposeEvent]: Purpose } = {
+    functional: {
+        state: null,
+        callbacks: [],
+    },
+    performance: {
+        state: null,
+        callbacks: [],
+    },
+    advertisement: {
+        state: null,
+        callbacks: [],
+    },
+};
+
+const triggerConsentNotification = (): void => {
+    Object.keys(purposes).forEach(key => {
+        const purpose = purposes[key];
+        purpose.callbacks.forEach(callback => callback(purpose.state));
+    });
+};
+
+export const onConsentNotification = (
+    purposeName: PurposeEvent,
+    callback: PurposeCallback
+): void => {
+    const purpose = purposes[purposeName];
+
+    if (cmpIsReady) {
+        callback(purpose.state);
+    }
+
+    purpose.callbacks.push(callback);
+};
+
+export const consentState = (purposeName: PurposeEvent): boolean | null =>
+    purposes[purposeName].state;
+
+export const init = (): void => {
+    // do nothing if init has already been called
+    if (cmpIsReady) {
+        return;
+    }
+
+    /**
+     * These state assignments are temporary
+     * and will eventually be replaced by values
+     * read from the CMP cookie.
+     * */
+    purposes.functional.state = true;
```

Comment:
Even just as temporary placeholders, I'm nervous about these starting `true`. Consent is `false` until explicitly given and it might be good to have us start thinking that way from the start, code-wise.

Reply:
The reason we've temporarily set the initial state of `functional`/`performance` is so we can do all the preparatory work to release this library (`lib/cmp.js`) and use it, without the CMP UI having to be in production collecting consent states. 

The site's code is currently executed as if the `functional`/`performance` consent states for all user's were `true`, because there's no consent state against these purposes to check against, as to date we've never asked. As soon as we turn on the switch that enables the CMP UI to begin collecting consent for these purposes we would stop using these placeholders and begin using the consent state from the cookie.

Would it be beneficial to add the switch into this PR, and when it's OFF we set these values to `true` (as we're doing now) and when it's ON we use the defaults we'll be working with going forward post CMP UI launch? Whatever they may be, `null` I presume...

Reply:
Technically there's no `defaultTrue`/`defaultFalse` states so I think it's  unnecessary to have this distinction with `userTrue`/`userFalse`. So the 3 potential values of consent state would be 'defaultNull`, `userTrue` and `userFalse`, It'd be good to understand the benefits of using these values as an ENUM as appose to `null`, `true`, `false`?

Reply:
@gtrufitt we've raised a separate PR for the module under discussion here https://github.com/guardian/frontend/pull/21705